File: java/java_tests/src/test/java/com/entrepidea/algo/tests/list/LinkedListUtils.java
package com.entrepidea.algo.tests.list;

import java.util.concurrent.ThreadLocalRandom;

/*
* A helper class designed to be used by other classes in this package
*
* */
public class LinkedListUtils {
    private static LinkedListUtils ourInstance = new LinkedListUtils();

    public static LinkedListUtils getInstance() {
        return ourInstance;
    }

    private LinkedListUtils() {
    }


    /***************************************************************
     *
     * Below are the help methods for this test
     *
     ****************************************************************/
    //double linked list node
    static class DListNode{
        int val;
        DListNode left;
        DListNode right;
        DListNode(int v){val = v; }
    }

    //singly linked list node
    static class ListNode {
        int val;
        ListNode next;
        ListNode(int x) { val = x; }
    }

    //construct a single linked list for testing
    public ListNode constructList(int len){
        ListNode head = new ListNode(0);
        ListNode p = head;
        for(int i=1;i<len;i++){
            ListNode node = new ListNode(i);
            p.next = node;
            p = p.next;
        }
        return head;
    }

    public ListNode constructRandomNumberList(int len, int numRange){
        int[] arr = ThreadLocalRandom.current().ints(0,numRange).limit(len).toArray();
        return constructListFromArray(arr);
    }


    public ListNode constructListFromArray(int[] arr){
        ListNode head = new ListNode(arr[0]);
        ListNode p = head;
        for(int i=1;i<arr.length;i++){
            ListNode node = new ListNode(arr[i]);
            p.next = node;
            p = p.next;
        }
        return head;
    }

    //print the list for testing
    public void printList(ListNode head){
        ListNode p = head;
        while(p!=null){
            System.out.print(p.val);
            System.out.print('\t');
            p = p.next;
        }
        System.out.println();
    }



    public ListNode findLastNode(ListNode head){
        ListNode p = head;
        while(p.next!=null){
            p = p.next;
        }
        return p;
    }
}

File: python/tutorials/algo/leetcode/easy/utils/LinkedListUtils.py
from typing import Optional

class ListNode:
    def __init__(self, data):
        self.data = data
        self.next = None
    """
    def __repr__(self):
        return f'C({self.data})'
    def __hash__(self):
        return hash(self.data)
    def __eq__(self,other):
        return self.__class__ == other.__class__ and self.data == other.data
    """

def construct_linked_list(arr) -> ListNode:
    head = ListNode(arr[0])
    node = head

    for i in range(1, len(arr)):
        node.next = ListNode(arr[i])
        node = node.next
    return head

def print_list(head : Optional[ListNode]):
    p = head
    while p is not None:
        print(p.data, end=',')
        p = p.next
    
    print()

def create_cycle(head : Optional[ListNode],pos) -> ListNode:
    p = head
    i =0
    node = None
    while p.next is not None:
        if i == pos:
            node = p
        p = p.next
        i+=1
    p.next = node
    return head

